### ISSUE TITLE
iOS Unity 2019.4 and UnityFramework pod target error

### ISSUE THREAD

**Issue by Str4tos** [googlesamples/unity-jar-resolver#405]
### [REQUIRED] Please fill in the following fields:

  * Unity editor version: **2019.4.11f1, 2020.1.7f1**
  * External Dependency Manager version: v1.2.159 
  * Source you installed EDM4U: Unity Package Manager
  * Features in External Dependency Manager in use: iOS Resolver
  * Plugins SDK in use: CleverAdsSolutions (Own iOS Framework)
  * Platform you are using the Unity editor on: Mac

### [REQUIRED] Please describe the question here:
Hello.  
I am using Cocaopods to integrate my own framework to Unity build.  
Everything works well on older versions of Unity like 2018 or 2019.2.  
However, in new versions of Unity, a new iOS Target Unity Framework was added. And at this moment problems appear.   

EDM4U creates the following Podfile:
```
source 'https://github.com/CocoaPods/Specs.git'
source 'https://github.com/cleveradssolutions/CAS-Specs.git'
platform :ios, '11.0'

target 'UnityFramework' do
  pod 'CleverAdsSolutions-Promo', '~> 1.6.1'
  pod 'CleverAdsSolutions-SDK', '~> 1.6.1'
end
```

The XCode project builds successfully, but an Fatal Error appears during application launch:
```
Error loading [path]/Test2020.app/Frameworks/UnityFramework.framework/UnityFramework:  
  dlopen([path]/Test2020.app/Frameworks/UnityFramework.framework/UnityFramework, 265): 
     Library not loaded: @rpath/CASPromo.framework/CASPromo
  Referenced from: [path]/Test2020.app/Frameworks/UnityFramework.framework/UnityFramework
  Reason: image not found
```

I see a solution how to fix this error: modify the Podfile as follows:
```
source 'https://github.com/CocoaPods/Specs.git'
source 'https://github.com/cleveradssolutions/CAS-Specs.git'
platform :ios, '11.0'

target 'UnityFramework' do
  pod 'CleverAdsSolutions-Promo', '~> 1.6.1'
  pod 'CleverAdsSolutions-SDK', '~> 1.6.1'
end

target 'Unity-iPhone' do
  pod 'CleverAdsSolutions-Promo', '~> 1.6.1'
  pod 'CleverAdsSolutions-SDK', '~> 1.6.1'
end
```

In this case, all errors go away. And there are no warnings about duplicate code in the build.  
I found a closed discussion after which it was decided to leave only one target in the podfile: #282.  

Please tell me how to fix it better.  
Perhaps I can somehow change the settings of my framework so that it always gets into the build. Or change the settings of Unity Project.  

Thank you very much for your attention.

**Comment by chkuang-g** [comment]
Hi @Str4tos 

Since you mentioned about EDM4U and iOS Resolver, I assume you have a `Dependencies.xml` in your Unity project so that EDM4U can generate the Podfile from, is that correct?

You solution seem to be actually opposite to what is mentioned in #282 so I am a bit confused.

Does it work if you only have `UnityFramework` and run `pod repo update; pod install`?

Otherwise, could you provide a minimum viable Unity project and steps to reproduce the issue?

Thanks
Shawn

**Comment by Str4tos** [comment]
Hi @chkuang-g 

> Since you mentioned about EDM4U and iOS Resolver, I assume you have a Dependencies.xml in your Unity project so that EDM4U can generate the Podfile from, is that correct?  

Yes, i have a `Dependencies.xml` in Unity Project and EDM4U generate the Prodfile only for UnityFramework.

> You solution seem to be actually opposite to what is mentioned in #282 so I am a bit confused.  

Yes, my solution is actually  opposite to what is mentioned in #282, but somehow my solution helps solve the problem.

> Does it work if you only have UnityFramework and run pod repo update; pod install?  

This does not work.  
To solve `Error loading` application, need to copy the dependencies for target 'Unity-iPhone' as I showed in the first post and run `pod install`.

> Otherwise, could you provide a minimum viable Unity project and steps to reproduce the issue?  

Of course, below will be the project archive for Unity 2019.4, but the same problem will be repeated for Unity 2020.1.
[CAS-UnitySample.zip](https://github.com/googlesamples/unity-jar-resolver/files/5414336/CAS-UnitySample.zip)

`Dependencies.xml` path: Assets/CleverAdsSolutions/Editor

Steps to reproduce the issue:
1. Open Unity project from archive
2. Build for iOS platform
3. Open XCode Project
4. Run XCode Project  

I do not deny that I may not know something in the framework settings. And I will be glad if you tell me what I need to do to solve such problems.

Many thanks

**Comment by polmum** [comment]
Hey!
This is also happening to us, it appears to be an issue with the dependency including a dynamic framework, which is not being copied by cocoapods as there is no host target defined in the podfile (when running `pod install`, the following warning appears: `The Podfile contains framework or static library targets (UnityFramework), for which the Podfile does not contain host targets (targets which embed the framework)`).
I found that just by adding the `Unity-iPhone` target to the Podfile, without any pod definitions, seems to fix this issue:
```
target 'Unity-iPhone' do

end
```
If this is indeed the case, I think this should be added to the podfile generation post process.

**Comment by vimanyu** [comment]
Thanks @polmum for listing that error. Looks like the app needs to be aware of its dependency on the framework.
@Str4tos, I tried reproducing this locally using your project but since I dont have an iOS device, I used simulator SDK and saw no errors while running the app in a simulator.
1. Could you confirm if you are seeing this error even on simulator? 
2. I havent tried this but just out of curiosity, could you also try this podfile that embedding the framework inside the Unity-IPhone target,

```
source 'https://github.com/CocoaPods/Specs.git'
source 'https://github.com/cleveradssolutions/CAS-Specs.git'
platform :ios, '11.0'

target 'UnityFramework' do
  pod 'CleverAdsSolutions-Promo', '~> 1.6.1'
  pod 'CleverAdsSolutions-SDK', '~> 1.6.1'
  target 'Unity-iPhone' do
  end
end
```

**Comment by Str4tos** [comment]
Thanks @vimanyu and @polmum 

I tried different options on Unity 2020.1.16f, Xcode 12.2 and here's what happened.
> pod 'CleverAdsSolutions-SDK/Base' - to test only one swift framework without other dependencies. 
> Application is removed before each test.  

1. Default unity build with following podfile:
```cpp
source 'https://github.com/CocoaPods/Specs.git'
source 'https://github.com/cleveradssolutions/CAS-Specs.git'
platform :ios, '11.0'

target 'UnityFramework' do
  pod 'CleverAdsSolutions-SDK/Base'
end
```
Device iOS 14: **I get an Error loading: Library not loaded**
Simulator iOS 14: **I get an Error loading: Library not loaded**

2. Unity build using the following podfile:
```cpp
source 'https://github.com/CocoaPods/Specs.git'
source 'https://github.com/cleveradssolutions/CAS-Specs.git'
platform :ios, '11.0'

target 'UnityFramework' do
  pod 'CleverAdsSolutions-SDK/Base'
  target 'Unity-iPhone' do
  end
end
```
Save and call: `pod install --no-repo-update`
Device iOS 14: **Works great**
Simulator iOS 14: **Works great**

3. Unity build using the following podfile:
```cpp
source 'https://github.com/CocoaPods/Specs.git'
source 'https://github.com/cleveradssolutions/CAS-Specs.git'
platform :ios, '11.0'

target 'UnityFramework' do
  pod 'CleverAdsSolutions-SDK/Base'
end

target 'Unity-iPhone' do
end
```
Save and call: `pod install --no-repo-update`
Device iOS 14: **Works great**
Simulator iOS 14: **Works great**

> Full pod 'CleverAdsSolutions-SDK' and 'CleverAdsSolutions-Promo' works same.

I believe that the third option is enough to solve the problem.

**Comment by vimanyu** [comment]
Great! Thanks for testing these combinations and I am flagging this as a bug just so we can do this automatically during the resolve.

**Comment by Str4tos** [comment]
@patm1987 could you please clarify what additional information you want?

I see that this is some kind of specific problem and this will not be a quick fix for EDM4U, so I have to write a Editor script to add lines to the podfile.
```c#
#if UNITY_2019_3_OR_NEWER
  [PostProcessBuild( 45 )]//must be between 40 and 50 to ensure that it's not overriden by Podfile generation (40) and that it's added before "pod install" (50)
  private static void FixPodFile( BuildTarget target, string buildPath )
  {
      using (StreamWriter sw = File.AppendText( buildPath + "/Podfile" ))
      {
          sw.WriteLine( "\ntarget 'Unity-iPhone' do\nend" );
      }
  }
#endif
```

**Comment by Str4tos** [comment]
I recently saw the same solution to the problem by [Facebook Unity SDK](https://github.com/facebook/facebook-sdk-for-unity/blob/6bb239bc4e10d9cc79f937245bf7b02c625785d1/Facebook.Unity.Editor/FacebookPostprocess.cs#L32-L50).

**Comment by vimanyu** [comment]
Thanks for posting more information. Your suggestion is valid and we will incorporate it into our next release of EDM4U.

**Comment by chkuang-g** [comment]
Hi folks,

In short, I think Facebook solution might not be the best one.

You can try to add `use_frameworks! :linkage => :static` in Podfile instead of an empty `Unity-iPhone` target.

I posted my findings here https://github.com/firebase/quickstart-unity/issues/862#issuecomment-758893827 . Please take a look when you have time.

I am planning to add this option to EDM4U.  

Please let us know if this works for you.
Shawn

**Comment by Str4tos** [comment]
Hi @chkuang-g 

I took the time to check your solution to the problem. 
- Unity 2021.1.16
- Cocaopods 1.10.1. 
- XCode 12.2.  
- macOS Catalina 10.15.4
- Target iOS 14.2

My framework is full swift 5 code packed in `.XCFramework`
I went ahead and started trying different `use_frameworks!` options.

1.  Option: 
However, your podfile **did not work for me**. 
I get an Error loading: Library not loaded
```cpp
source 'https://github.com/CocoaPods/Specs.git'
source 'https://github.com/cleveradssolutions/CAS-Specs.git'
platform :ios, '11.0'

target 'UnityFramework' do
  pod 'CleverAdsSolutions-SDK/Base'
end

use_frameworks! :linkage => :static
```

2.  Option:
Unfortunately I found that my framework won't load even with the Facebook solution. 
With `use_frameworks! :linkage => :static` same load error.
But if you remove `use_frameworks!`, everything works fine.
```cpp
source 'https://github.com/CocoaPods/Specs.git'
source 'https://github.com/cleveradssolutions/CAS-Specs.git'
platform :ios, '11.0'

use_frameworks!

target 'UnityFramework' do
  pod 'CleverAdsSolutions-SDK/Base'
end

target 'Unity-iPhone' do
end
```

3.  Option:
I get an Error loading: Library not loaded
```cpp
source 'https://github.com/CocoaPods/Specs.git'
source 'https://github.com/cleveradssolutions/CAS-Specs.git'
platform :ios, '11.0'

target 'UnityFramework' do
  use_frameworks! :linkage => :static
  pod 'CleverAdsSolutions-SDK/Base'
end
```

4.  Option:
**Works great** with `use_frameworks!` and `use_frameworks! :linkage => :static` 
```cpp
source 'https://github.com/CocoaPods/Specs.git'
source 'https://github.com/cleveradssolutions/CAS-Specs.git'
platform :ios, '11.0'

use_frameworks! :linkage => :static

target 'UnityFramework' do
  pod 'CleverAdsSolutions-SDK/Base'
end

target 'Unity-iPhone' do
  pod 'CleverAdsSolutions-SDK/Base'
end
```
At the same time, there are no messages about code duplication during application launch. 
Although I have tried the full package that includes a lot of well-known ad networks.

**Comment by polmum** [comment]
@chkuang-g 
I've seen [here](https://github.com/facebook/facebook-sdk-for-unity/issues/510) that you are planning to add fixes to EDM4U to address this issue.

Could you maybe share the proposed fix? I'm asking because I think that adding `use_frameworks!` to the Podfile should always be optional, so any possible solution should consider a Podfile with and without this line added.

**Comment by chkuang-g** [comment]
@patm1987 

After investigation, in order to support Swift libraries, adding both `use_frameworks!` and an empty `Unity-iPhone` target to Podfile seems the only configuration that works for the case for CleverAdsSolutions and Facebook SDK. That is, adding the following lines to Podfile.
```
use_frameworks!
target 'Unity-iPhone' do
end
```

If you are using Firebase, you also need to make sure `Google-Services.plist` is added to `Unity-iPhone` target membership.  You can use https://github.com/firebase/quickstart-unity/issues/862#issuecomment-752945417 to solve this from Unity.

For Podfile, I'm planning to add two options to EDM4U.
<img width="399" alt="Screen Shot 2021-01-26 at 10 12 35 PM" src="https://user-images.githubusercontent.com/31869252/106040746-b8af0600-608f-11eb-96c4-f38efb7c5134.png">

And also patch Firebase editor tool to add `Google-Services.plist` to all available target, except test target.

Shawn

**Comment by polmum** [comment]
@chkuang-g
Adding only the empty `Unity-iPhone`, without `use_frameworks!`, should also fix the issue. At least in our case this is our currently working solution.
That being said, I think adding two new separate options to EDM4U is a good call.

**Comment by chkuang-g** [comment]
https://github.com/googlesamples/unity-jar-resolver/commit/256994f1b51d5a8d7c7679a77926fdb9eb40e037
https://github.com/googlesamples/unity-jar-resolver/commit/25e13efbce88bf34dea6c853376a2960e0594f7f

This change should give you enough options. They are released in `1.2.164`.

However, it is really odd that your team (CleverAdsSolutions) need to duplicate pods in all targets.  This will cause link error if the user choose to statically link everything.  I would recommend you to dig a bit more on that.  This may bring pain your users when they need to use other 3rd party libraries.
